Draft: Add site language switcher to gutenboarding

roo2 opened the pull request:
Build on top of the work that @sirreal did in #39627 this PR allows live switching of the language in gutenboarding, and adds a language picker made from core components.

I changed from using an actual modal component to using a step because that is now how the domains and plans steps have been implemented.

I've got it behind a feature flag for now because I'd like to get the language switching changes and refactoring tested and merged before the UI is completely polished.

### Functionality
Can now change the site language dynamically in gutenboarding using the `window.updateLocale('en')` function

### Changes
* Wraps the app in a new LocaleContext component which exposes a function for changing the locale.
* Breaks the initial locale loading code out of gutenboarding/index.tsx and into `locale-context.tsx`.
* Handles chunked translations as well as non chunked translations
* Includes the actual language picker now as a modal step
* The language picker doesn't currently have search functionality, I'd like to implement that as an additional PR because this one is already big! 

### Limitations
* Does not set the language slug in the URL, language gets lost on refresh
* Does not set the language in the users settings after completing the flow
* Search for languages is not implemented

Part of https://github.com/Automattic/wp-calypso/issues/40011

### Testing instructions

* This is hidden behind a feature flag, enable it through the URL by appending `?flags=gutenboarding/language-picker`
* The <HTML> Element should have a `lang` attribute set appropriately for the current language
* The html element should have its `dir` attribute set to `rtl` for right to left languages such as arabic
* The language picker step should behave similarly to the plans and domain picker steps
* In production, we use Chunked Translations which is a different way of loading translations, affected by this PR. 
    * Chunked translations can be tested on ci by adding `?flags=gutenboarding/language-picker,` 
    * Or tested locally by running the following command `ENTRY_LIMIT=entry-gutenboarding BUILD_TRANSLATION_CHUNKS=true ENABLE_FEATURES=use-translation-chunks,gutenboarding/language-picker yarn start`

<img width="1187" alt="Screen Shot 2020-07-14 at 7 16 46 PM" src="https://user-images.githubusercontent.com/22446385/87409249-0be7cc80-c608-11ea-9157-0f9245a41eaf.png">

fixes #40011

roo2 commented on client/landing/gutenboarding/index.tsx:
```diff
@@ -94,46 +130,36 @@ window.AppBoot = async () => {
+		switchWebpackCSS( rtl );
+	}, [ rtl ] );
+	return null;
+}
```
hey, it looked like the result of that code would just be to also add the lang attribute to the <html/> element like `<html lang="ar" dir="rtl">` I copied the relevant bit of https://github.com/Automattic/wp-calypso/blob/09cbaab35dd1b5e1e01720e4771b05974c5b2939/client/lib/i18n-utils/switch-locale.js#L93-L101
and just used `i18nLocale` from the `react-i18n` package to get the `lang`

roo2 commented on packages/react-i18n/src/index.tsx:
```diff
@@ -16,17 +16,51 @@ export interface I18nReact {
 }
-export const I18nProvider: React.FunctionComponent< Props > = ( { children, localeData } ) => {
+
+export const I18nProvider: React.FunctionComponent< Props > = ( {
```
@sirreal does this make sense to you to expand this I18nProvider class to also include changing locales? It required me to pass in some extra parameters, especially the `getLanguageFile` function. I was thinking it might even make sense to share this with calypso eventually? related to #42997

sirreal replied:
This might make sense eventually, but I'd like to leave this package as is and implement locale changing in Calypso application code. The package is quite simple and focused on providing the translation functionality, I'm reluctant to have it start managing more things.

roo2 replied:
ok, that makes sense 👍

sirreal commented on client/landing/gutenboarding/index.tsx:
```diff
@@ -134,6 +131,20 @@ window.AppBoot = async () => {
+	const isRtl = isRTL();
+
+	React.useEffect( () => {
+		document.documentElement.lang = i18nLocale;
```
This should be in its own effect, it may change independently of RTL.

Let's get the lang like this:

https://github.com/Automattic/wp-calypso/blob/1826519f53921884b1dcd03d0ad79cbdc52403a3/client/lib/i18n-utils/switch-locale.js#L79-L91

In our case, I'd expect:

```ts
const { __, isRTL, i18nLocale } = useI18n();
const slug = __( 'html_lang_attribute' );
// …
```

roo2 replied:
thanks oh I see that the `html_lang_attribute` is different from the slug, oops. I've made those changes now.

simison commented on client/landing/gutenboarding/components/locale-context/index.tsx:
```diff
@@ -0,0 +1,48 @@
+/**
```
Would there be better, more generic place for this component than in `client/landing/gutenboarding/components/` ?

cc @akirk for thoughts.

simison commented on client/landing/gutenboarding/index.tsx:
```diff
@@ -129,11 +122,38 @@ window.AppBoot = async () => {
 	);
 };
 
+function WindowLocaleEffectManager() {
```
Similarly to https://github.com/Automattic/wp-calypso/pull/43132/files#r444074316 this function could do with more re-usable friendly location.

roo2 replied:
Thanks, I pulled that out to a separate component 👍

roo2 commented on client/landing/gutenboarding/index.tsx:
```diff
@@ -94,31 +95,23 @@ window.AppBoot = async () => {
-			| any
-		 )[] = await getLocale();
-		i18nLocaleData = localeData;
-		setLocaleData( localeData );
```
I removed setLocaleData from here because it is called in https://github.com/WordPress/gutenberg/blob/master/packages/i18n/src/create-i18n.js#L189 via the `<I18nProvider>`https://github.com/Automattic/wp-calypso/blob/master/packages/react-i18n/src/index.tsx#L72

roo2 commented:
@sirreal If you get time, this is ready for another look now, I moved the changeLocale function into a provider and its own component, fixed the RTL and html lang attribute code and pass the `lang_id` along to the `create_site` function.

Thanks!

roo2 commented:
@razvanpapadopol this implements changing language like we touched on in https://github.com/Automattic/wp-calypso/pull/43609 if you have any spare cycles, I think you'd be a good person to take a look at this 🙏

sirreal commented:
To be clear and transparent: Reviewing this is in my queue, but I'm swamped. It would be great if other folks could review this. I'm not sure when I'll be able to get back to this.

yuliyan commented on client/landing/gutenboarding/components/window-locale-effect-manager/index.tsx:
```diff
@@ -0,0 +1,33 @@
+	React.useEffect( () => {
+		document.documentElement.lang = lang;
+	}, [ lang ] );
+
```
```suggestion
	// Ensure `ltr` string will be extracted into calypso-strings.pot
	// Since string extraction is __ calls as if they were i18n-calypso's translate calls,
	// `context` must be passed the same way.
	__( 'ltr', { context: 'text direction' } );
```

I discovered that `isRTL` wasn't working with translation chunks and it's related to how [`@wordpress/i18n` implements it](https://github.com/WordPress/gutenberg/blob/master/packages/i18n/src/create-i18n.js#L185). Internally, it's calling gettext with `ltr` original string, but since it's not part of Calypso's monorepo, it ends up not being included in `calypso-strings.pot`, and therefore not being included in Gutenboarding's generated translation chunks. As a workaround, we can simply call the gettext to ensure `ltr` will be extracted as original string from the repo.

There's another catch here - we're are still using `i18n-calypso` to extract strings ([switching to alternative workflow is in the progress](https://github.com/Automattic/wp-calypso/pull/39541)) so `__` calls are actually being handled as if they were `translate` calls from `i18n-calypso`, and therefore `context` must be passed the same way.

roo2 replied:
Thanks for the investigation and finding this bug!

hmmm I still don't understand this, I tested the code `__( 'ltr', { context: 'text direction' } );` locally but  for arabic it just returned `ltr` instead of `rtl` 🤔 .

I was able to inspect the contents of the file `./calypso-strings.pot` and yep, `ltr` wasn't in it.
I did find `rtl` tucked away inside `localeData[''].momentjs_locale.textDirection`, it might be possible to use that? https://github.com/Automattic/wp-calypso/pull/43132/files#diff-2d167e48666aab3480ccd61930e3cb7aR55

But hmmm I still can't understand why the original code or your suggestion is not working, do you have any more insights or could you explain again your understanding of what's going on here? 🙏 

cc @akirk I've heard you also have insights on i18n stuff ;)

yuliyan replied:
If you've added `__( 'ltr', { context: 'text direction' } );` after dev process has been started, it wouldn't work `calypso-strings.pot` and translation chunks are only being generated once on boot.

If that's the case, you can either restart the dev process, or run `yarn run build-languages` in a separate terminal window.

roo2 replied:
ahhhh I spent some time chasing my tail on this but I now understand, just adding that line of text will cause the `calypso-i18n` tool to include a line in `./calypso-strings.pot`, which will then allow the `./bin/build-languages` script to include an entry in `ar-entry-gutenboarding.js` like so `"text direction\u0004ltr":["rtl"]`. which can then be accessed by https://github.com/WordPress/gutenberg/blob/master/packages/i18n/src/create-i18n.js#L185. 

One thing that confused me, is that using the double underscore translation method `__( 'ltr', { context: 'text direction' } );` does not actually pass the context correctly to `dcnpgettext` 🤔. the `__` method defines the second argument as `domain` https://github.com/WordPress/gutenberg/blob/master/packages/i18n/src/create-i18n.js#L118. The `_x` function does pass context as the second argument. The `_x` function is used internally by isRTL https://github.com/WordPress/gutenberg/blob/master/packages/i18n/src/create-i18n.js#L185

But when I replaced `__( 'ltr', { context: 'text direction' } );` with `_x` (`_x( 'ltr', 'text direction' );`) and ran `yarn run build-languages`, the translation for ltr was not included in the `ar-entry-gutenboarding.js` file. This would seem to be a limitation of the `i18n-calypso` cli tool which is called as part of `yarn run build-languages` and is responsible for generating the `./calypso-strings.pot` file.

yuliyan replied:
> But when I replaced __( 'ltr', { context: 'text direction' } ); with _x (_x( 'ltr', 'text direction' );) and ran yarn run build-languages, the translation for ltr was not included in the ./calypso-strings.pot or ar-entry-gutenboarding.json files. This would seem to be a limitation of the i18n-calypso cli tool which is called as part of yarn run build-languages and is responsible for generating the ./calypso-strings.pot file.

That's because `__`, `_x`, `_n`, `_nx` are used as aliases for `translate` and therefore it must be called with the same parameters format to get it extracted into `calypso-strings.pot` properly. It won't get translated properly, but since we'll only be using it to extract the string into the POT file, it should be fine for now.

We plan on switching to a different string extraction process. https://github.com/Automattic/wp-calypso/pull/39541 has been open for quite some time now, but I'll be getting back working on getting it done. It's using [wp-babel-makepot](https://github.com/Automattic/wp-babel-makepot) which is based on [babel-plugin-i18n-calypso](https://github.com/Automattic/wp-calypso/tree/master/packages/babel-plugin-i18n-calypso) and it should be handling `translate`, `__`, `_x`, `_n` and `_nx` all individually.

sirreal replied:
If I understand everything, we need to trick the extraction into including the `ltr` with context `text direction` translation into the translation.

The correct way to call a context translation with `@wordpress/i18n` is `_x( 'ltr', 'text direction' )`, or the helper `isRTL`.

Below we still use `isRTL` to fetch the localized string. Would this creating and invoking a translation function to be picked up by the extraction have the same effect?

```ts
// We need to invoke translate in this way so that the `ltr` localization is included in the localization bundle.
const translate = () => undefined;
translate( 'ltr', { context: 'text direction' } );
```

yuliyan replied:
I haven't tested your suggestion, but it should work since it's [only matching function name](https://github.com/Automattic/xgettext-js/blob/master/xgettext.js#L214-L216).

roo2 commented on client/landing/gutenboarding/components/locale-context/index.tsx:
```diff
@@ -0,0 +1,203 @@
+
+			if ( USE_TRANSLATION_CHUNKS ) {
+				const chunkedLocaleData = await setupTranslationChunks( newLocale, translatedChunks );
+				setLocale( chunkedLocaleData );
```
thank you, good catch!

roo2 commented on client/landing/gutenboarding/components/locale-context/index.tsx:
```diff
@@ -0,0 +1,203 @@
+const DEFAULT_LOCALE_SLUG: string = config( 'i18n_default_locale_slug' );
+const USE_TRANSLATION_CHUNKS: boolean =
+	config.isEnabled( 'use-translation-chunks' ) ||
+	getUrlParts( document.location.href ).searchParams.has( 'USE_TRASNSLATION_CHUNKS' );
```
ahh good catch 👍

simison commented on client/landing/gutenboarding/onboarding-block/language/index.tsx:
```diff
@@ -58,13 +61,11 @@ const LanguageStep: React.FunctionComponent = () => {
-				</Button>
+				<div key={ languageGroup.id }>
+					<Button onClick={ onClick } className={ 'language__language-group' }>
+						<span className={ isSelected ? 'selected' : '' }>{ languageGroup.name( __ ) }</span>
```
Would be good to have `is-selected` format for modifier classes.
https://wpcalypso.wordpress.com/devdocs/docs/coding-guidelines/css.md?term=css

roo2 replied:
thanks I updated that 👍

ramonjd commented on client/landing/gutenboarding/onboarding-block/language/index.tsx:
```diff
@@ -0,0 +1,107 @@
+									className="language__language-item"
+									key={ language.langSlug }
+									onClick={ () => {
+										changeLocale( language.langSlug );
```
Just a question: would it make sense in a future iteration to append the language slug to the url as well? I.e., `wordpress.com/new/it` 

Just thinking if the page refreshes for whatever reason.

roo2 replied:
Hmmm yes, currently when you refresh the page, the selected language is lost. With the implementation of the modal step, I use history.back() to close out of the step. I can't think of a good way to close out of the step while also adding the language slug to the URL.

I'd like to leave this as un-implemented for now

ramonjd commented:
Functionally, this tests really well for me. Language switching works smoothly, including RTL. 

Kudos to @roo2 for going the long and hard yards with this one. 🙇 

It's behind a feature flag so do folks have any qualms about merging this branch so we can iterate on the UI?

Aside from the small class name convention that @simison [notes](https://github.com/Automattic/wp-calypso/pull/43132/files#r454163694), it might be handy to add the language slug to the URL after switching.

Also for some future iteration, we're not passing the `locale` param to rest/v1.5/plans in https://github.com/Automattic/wp-calypso/blob/599c098/packages/data-stores/src/plans/resolvers.ts#L45, e.g., https://public-api.wordpress.com/rest/v1.5/plans?locale=it

ramonjd commented:
I'm just looking at things after the latest commit and now I can only see the language picker on http://calypso.localhost:3000/new/design and for none of the previous steps.

Am I doing something wrong? Just compiled locally using `BUILD_TRANSLATION_CHUNKS=true ENABLE_FEATURES=use-translation-chunks,gutenboarding/language-picker yarn start` but I noticed it on calypso.live/new too.

roo2 commented:
> I'm just looking at things after the latest commit and now I can only see the language picker on http://calypso.localhost:3000/new/design and for none of the previous steps.

Ahhh yes, good catch, I had hooked into the same logic that Plans and Domains steps use for hiding the header buttons when the user is in a "Modal" step. But yea that didn't actually make sense for the language picker, you want it to always be available, and certainly on the homepage. I pushed a fix, thanks.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/4110900

Thank you @roo2 for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/4110900).
